source/MessageLogPanel.cpp

```diff
@@ -0,0 +1,148 @@
+/* MessageLogPanel.cpp
+Copyright (c) 2023 by TomGoodIdea
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "MessageLogPanel.h"
+
+#include "text/alignment.hpp"
+#include "Color.h"
+#include "Command.h"
+#include "FillShader.h"
+#include "text/Font.h"
+#include "text/FontSet.h"
+#include "GameData.h"
+#include "Interface.h"
+#include "Preferences.h"
+#include "Screen.h"
+#include "Sprite.h"
+#include "SpriteSet.h"
+#include "SpriteShader.h"
+#include "UI.h"
+#include "text/WrappedText.h"
+
+using namespace std;
+
+namespace {
+	const double PAD = 10.;
+	const double LINE_HEIGHT = 25.;
+}
+
+
+
+MessageLogPanel::MessageLogPanel(const deque<pair<string, Messages::Importance>> &messages)
+	: messages(messages)
+{
+	SetInterruptible(false);
+}
+
+
+
+void MessageLogPanel::Draw()
+{
+	// Dim out everything outside this panel.
+	DrawBackdrop();
+
+	// Draw the panel.
+	width = Screen::Width() - GameData::Interfaces().Get("message log")->GetValue("margin");
+	const Color &backColor = *GameData::Colors().Get("message log background");
+	FillShader::Fill(
+		Point(Screen::Left() + .5 * width, 0.),
+		Point(width, Screen::Height()),
+		backColor);
+
+	const Sprite *edgeSprite = SpriteSet::Get("ui/right edge");
+	if(edgeSprite->Height())
+	{
+		// If the screen is high enough, the edge sprite should repeat.
+		double spriteHeight = edgeSprite->Height();
+		Point pos(
+			Screen::Left() + width + .5 * edgeSprite->Width(),
+			Screen::Top() + .5 * spriteHeight);
+		for( ; pos.Y() - .5 * spriteHeight < Screen::Bottom(); pos.Y() += spriteHeight)
+			SpriteShader::Draw(edgeSprite, pos);
+	}
```

Comment:
Since this code is also in ConversationPanel, could we get a static method in a separate class that they both call?

Reply:
MapSalesPanel and LogbookPanel do this too.
https://github.com/endless-sky/endless-sky/blob/cf204d5424605de76c6b2c8d25e491b025636168/source/MapSalesPanel.cpp#L269-L280

https://github.com/endless-sky/endless-sky/blob/cf204d5424605de76c6b2c8d25e491b025636168/source/LogbookPanel.cpp#L95-L105

Reply:
Panel would make the most sense.

Reply:
Yeah, I'd be fine with Panel.